### DISCUSSION TITLE
Implementing divergence term in coupled variable

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#25448] [Q&A General] [answered]
Hello
I am trying to write the return expression for computeQpResidual() 
My expression is: `∇.(∇φ c) `
where c is the variable and φ is the coupled variable. Please verify whether I can write this as
My expression expands to:
`∇φ.∇c + c ∇.∇φ `

`_grad_phi[_qp] * _u[_qp] * _grad_test[_i][_qp]  + 2nd term`

The 2nd term has divergence of gradient of coupled variable. How, do I implement this:
Thanks

**Comment by GiudGiud** [comment]
Hello

On a term like this we typically use an integration by parts of the product of the term by the test function, rather than expand. 
We simply do not want to represent `∇.∇φ` in the finite element problem, as you need a second order basis to make that non-zero

Guillaume

**Reply by vermaprk** [reply]
Integration by parts followed by divergence results in weak form as:
`(∇Ψ,c∇φ) - <Ψ, c∇φ . n>` (please verify)
But, this results in coupled variable in boundary condition.
Thanks

**Reply by GiudGiud** [reply]
> But, this results in coupled variable in boundary condition.

That s not a problem, you can create a new one or just use Neumann if you know the flux

**Reply by vermaprk** [reply]
ok. Thanks Will trouble you again in writing the jacobians

**Reply by GiudGiud** [reply]
use automatic differentiation (AD) to avoid having to do that manually

**Comment by vermaprk** [comment]
> Integration by parts followed by divergence results in weak form as: `(∇Ψ,c∇φ) - <Ψ, c∇φ . n>` (please verify) But, this results in coupled variable in boundary condition. Thanks
Please check the maths:

```
∇.(Ψ(c∇φ)) = Ψ∇.(c∇φ) + c∇φ.∇Ψ
Ψ∇.(c∇φ) = ∇.(Ψ(c∇φ)) - c∇φ.∇Ψ
taking volume integral and using divergence theorem
-int (Ψ∇.(c∇φ)) dV = -int (Ψ(c∇φ.n))dS + int( c∇φ.∇Ψ) dv
weak form.              =  - <Ψ, c∇φ . n>      +  (∇Ψ,c∇φ) 
```

kernel coded as:
```
{
  return   _grad_test[_i][_qp] * _u[_qp] * _grad_phi[_qp];
}
```

However, I have found a different math used in code at:
https://github.com/anilkunwar/danphe/blob/working_codes/src/kernel/NernstPlanckConvection.C

line 63
```
{
return ion_velocity * _grad_u[_qp] * _test[_i][_qp];
}
```

Here grad of _u[_qp] is taken instead of the test function. 
Thanks

**Reply by vermaprk** [reply]
@GiudGiud

**Reply by GiudGiud** [reply, marked as answer]
Hello

That kernel you linked to is a convection term, it does not look like a diffusion term

If you look at a diffusion kernel you ll see what it should look like
https://github.com/idaholab/moose/blob/next/framework/src/kernels/Diffusion.C

which btw your diffusion term seems atypical? div c grad phi would mean that phi is the variable being diffused and c is the diffusivity?
Or am I missing something

the problem in the code you wrote is that 'grad_phi' corresponds to the gradient of the trial function.
It's not the gradient of the coupled variable. You need to do in your constructor:
```
... (rest of initialization list),
grad_coupled_var(coupledGradient(" parameter name for that coupled variable ")
```
and in the parameters you must have "variable = c" to make sure that _u corresponds to c. 

Then you can use for the residual:
```
ADReal
computeQpResidual
{
  return   _grad_test[_i][_qp] * _u[_qp] * _grad_coupled_var[_qp];
}
```

**Reply by vermaprk** [reply]
Yes, I have done the same thing but what I missed was that "phi is reserved for trial function". My coupled variable was also named as phi and the gradient of coupled variable was _grad_phi. However, I have benchmarked the results with phi being used as coupled variables with comsol. May be I have overwritten the trial function with my coupled variable.

I think its better to used some other name than phi.
Thanks
